contracts/Referendum.sol

```diff
@@ -27,82 +35,181 @@ contract Referendum {
+
+        ProposalOrigin _origin = ProposalOrigin.Proposal;
+        if (msg.sender == concilAddr) {
+            _origin = ProposalOrigin.Concil;
+        } else if (msg.sender == proposalQueueAddr) {
+            _origin = ProposalOrigin.ProposalQueue;
+        }
+
+        Proposal memory p = Proposal(_id, _origin, _prosFromConcil, false, new address[](0));
         proposals.push(p);
+        emit NewProposal(_id, _origin, _prosFromConcil);
+    }
+
+    /// @notice vote for proposal
+    /// @param _id id for the proposal to vote
+    /// @param _vType vote type, pros, cons or abs
+    function voteForProposal(uint _id, VoteType _vType) external proposalShouldExist(_id) returns (bool success) {
+        uint idx = getProposalIndex(_id);
+        Proposal storage p = proposals[idx];
+
+        for (uint vIdx = 0; vIdx < p.voters.length; vIdx++) {
+            if (p.voters[vIdx] == msg.sender) {
+                break;
+            }
+        }
+
+        if (vIdx == p.voters.length) {
+            p.voters.push(msg.sender);
+        }
+
+        p.votes[msg.sender] = _vType;
+        emit NewVote(_id, _vType, msg.sender);
+        return true;
     }
 
-    function checkProposal(uint _id) public view returns (uint _pros, uint _cons) {
-        Proposal storage p = proposals[_id];
+    /// @notice get votes of proposal by proposal id
+    /// @param _id proposal id
+    /// @return (uint pros, uint cons, uint abs)
+    function getVotesOfProposalById(uint _id) 
+    public 
+    proposalShouldExist(_id) 
+    view 
+    returns (uint pros, uint cons, uint abs) 
+    {
+        uint idx = getProposalIndex(_id);
+        Proposal storage p = proposals[idx];
+        for (uint i = 0; i < p.voters.length; i++) {
+            VoteType v = p.votes[p.voters[i]];
+            if (v == VoteType.Pros) {
+                pros = pros.add(1);
+            } else if (v == VoteType.Cons) {
+                cons = cons.add(1);
+            } else {
+                abs = abs.add(1);
+            }
+        }
+    }
+
+    /// @notice check proposal votes
+    /// if proposal is from concil with 100 prosFromConcil
+    ///    1. all abs are regarded as pros
+    ///    2. > 50% pros makes the proposal passed
+    ///
+    /// if proposal is from concil with >50 prosFromConcil
+    ///    1. divide abs into two parts according to pros/cons
+    ///    2. > 50% pros makes the proposal passed
+    /// 
+    /// if proposal is from proposal queue
+    ///    1. all abs are regarded as cons
+    ///    2. > 50% pros makes the proposal passed
+    ///
+    /// if proposal is from parent proposal
+    ///    1. inherits its parent proposal's standard
+    function checkProposalById(uint _id) public returns (bool accepted) {
+        // Proposal storage p = proposals[_id];
+        uint idx = getProposalIndex(_id);
+        Proposal storage p = proposals[idx];
+        (uint pros, uint cons, uint abs) = getVotesOfProposalById(_id);
```

Comment:
already have the index of the id, why use the getVotesOfProposalById instead of  get the these values directly?

Reply:
`getVotesOfProposalById` returns sums of `pros`, `cons`, `abs`, which is a mapping in `Proposal`

Reply:
i mean the `getProposalIndex` is a duplicate code, you call it at `getVotesOfProposalById` again

Reply:
Do you mean that the`getVotesOfProposalById` is redundant because `Proposal storage p = proposals[idx]` returns the `proposal` that provides all data to calculate `pros`, `cons`, `abs`? But `getVotesOfProposalById` is necessary for external call, if here I use a loop to get `pros` and more, there'll be two blocks to get `pros` of a `proposal`

Reply:
I mean you can define a private function: `getVotesOfProposalByIndex` or something like this

Reply:
that's ok.